### PR TITLE
Adding solve procedure with LAPACK support

### PR DESCRIPTION
**Pull request by CaptainSharf**
As part of this PR:
1. Adding the `solve ` routine to solve well determined linear equations of the form `Ax=B` where `A` is a square matrix, and the leading dimension of A matches with the number of rows in B
2. Adding tests for the solve procedure with LAPACK support

### CODE REVIEW

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -1500,14 +1500,58 @@ proc solve_triu (const ref U: [?Udom] ?eltType, const ref b: [?bdom] eltType) {
...
-proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
+proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) where !usingLAPACK{
   var (LU, ipiv) = lu(A);
   b = permute (ipiv, b, true);
   var z = solve_tril(LU, b);
   var x = solve_triu(LU, z);
   return x;
 }
 
+proc solve(A: [?Adom], B: [?Bdom]) where usingLAPACK{
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl:1511]: Example of style variation from rest of module:

```suggestion
proc solve(A: [?Adom], B: [?Bdom]) where usingLAPACK {
```

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -1500,14 +1500,58 @@ proc solve_triu (const ref U: [?Udom] ?eltType, const ref b: [?bdom] eltType) {
...
+  var hasError : int;
+  if(isSquare(A)){
+    if(Bdom.rank==1){
+      var rowsB = Bdom.size;
+      var B_mat :[1..rowsB,1..1] real;
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl:1516]: Example of style variation from rest of module:

```suggestion
      var B_mat :[1..rowsB, 1..1] real;
```

File: modules/packages/LinearAlgebra.chpl [hunk 3]
```diff
@@ -1500,14 +1500,58 @@ proc solve_triu (const ref U: [?Udom] ?eltType, const ref b: [?bdom] eltType) {
...
   var (LU, ipiv) = lu(A);
   b = permute (ipiv, b, true);
   var z = solve_tril(LU, b);
   var x = solve_triu(LU, z);
   return x;
 }
 
+proc solve(A: [?Adom], B: [?Bdom]) where usingLAPACK{
+  var hasError : int;
+  if(isSquare(A)){
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl:1513]: Example of style variation from rest of module:

```suggestion
  if isSquare(A) {
```

File: modules/packages/LinearAlgebra.chpl [hunk 4]
```diff
@@ -1500,14 +1500,58 @@ proc solve_triu (const ref U: [?Udom] ?eltType, const ref b: [?bdom] eltType) {
...
+    if(Bdom.rank==1){
+      var rowsB = Bdom.size;
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl:1515]: Example where we could use const:

```suggestion
      const rowsB = Bdom.size;
```

**Review by ben-albrecht — CHANGES_REQUESTED** [review]
Thanks for contributing @CaptainSharf! Most comments made inline.

Regarding testing, I believe `performance/lu/lu-solve-test.chpl` will now default to the LAPACK version. In order to preserve that performance data, let's add a `performance/lu/lu-solve-test.compopts` that disables LAPACK, e.g. `-slapackImpl=off`. 

We also want to test the LAPACK version. I think we can do that by adding another entry to the `perfcompopts` file with `-slapackImpl=lapack`. I am not sure how that interacts with the existing `perfexecopts`, however. We will also need to add lapack variant to the `performance/linearalgebra-perf.graph`.

Give it a try and let me know if you run into any issues or have questions.

Minor comments:

- I suggest using `const` when a variable is not expected to change
- Try to make your coding style match the rest of the module
    - I mostly see inconsistencies with spacing. I've pointed out a few examples inline.

**Comment by CaptainSharf** [comment]
Hi @bradcray ,
Thanks for the feedback! I'm trying to write performance tests for this method as you pointed out. Could you please point me towards how to execute performance tests, and verify the graphs?

**Comment by ben-albrecht** [comment]
>  Could you please point me towards how to execute performance tests, and verify the graphs?

Run `start_test --performance lu-solve-test.chpl` and then `start_test --gen_graphs`. It will print out a path which will contain an HTML file you can view in your browser.

**Comment by ben-albrecht** [comment]
The [test docs](https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst#a-performance-test) are a good reference as well.

**Comment by ben-albrecht** [comment]
@CaptainSharf - I am revisiting some old open PRs. I think this is a great feature to have, but I am going to close this PR due to inactivity for now. We'll reopen when you get a chance to resume development.

**Comment by CaptainSharf** [comment]
Thanks @ben-albrecht !
I wish to revisit this feature at some point in the near future. This PR is missing performance tests

**Pull request closed without merging**